### PR TITLE
TST: Nan must not be converted to string using .loc

### LINKED ISSUE #28403: NaN is converted to strings when reassigning a column with .loc
Hello team,

Here is a simple working exemple of value assignment on a dataframe:
```
df = pd.DataFrame({'A': [np.nan, np.nan, 'b']})
df['A'].loc[[0, 1]] = ['a', np.nan]
df.notna()
Out[126]: 
       A
0   True
1  False
2   True
```

However, if we use loc to assign the entire column with an array including NaNs, all values are converted to strings, breaking notna():
```
df = pd.DataFrame({'A': [np.nan, np.nan, 'b']})
df['A'].loc[[0, 1, 2]] = ['a', np.nan, np.nan]
df['A'].notna()                                                                                                                                                                                   
Out[133]: 
0    True
1    True
2    True
Name: A, dtype: bool
```

### PR DESCRIPTION
**Pull request by NumberPiOso**
- [X] Closes #28403
- [X] tests added / passed
- [X] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them

### CODE REVIEW

File: pandas/tests/indexing/test_loc.py [hunk 1]
```diff
@@ -43,6 +43,15 @@
 from pandas.tests.indexing.common import Base
 
 
+def test_not_change_nan_loc():
```
Comment (mroeschke) [inline pandas/tests/indexing/test_loc.py:46]: Could you also add a separate test function for the `df = pd.DataFrame({'A': [np.nan, np.nan, 'b']})` case in the issue as well?

Comment (NumberPiOso) [reply pandas/tests/indexing/test_loc.py:55]: Done

Comment (mroeschke) [reply pandas/tests/indexing/test_loc.py:55]: Could you use `pytest.mark.parameterize` to turn this into 1 test?

Comment (NumberPiOso) [reply pandas/tests/indexing/test_loc.py]: yes

**Comment by NumberPiOso** [comment]
ci checks are failing,
however error seems unrelated

```
Run ci/code_checks.sh typing
mypy --version
mypy 0.931
Performing static analysis using mypy
pandas/core/internals/blocks.py:1142: error: Argument 1 to "np_can_hold_element" has incompatible type "Union[dtype[Any], ExtensionDtype]"; expected "dtype[Any]"  [arg-type]
```

Running mypy at the file confirm the error, and the CI yields an even more errors.
```
> mypy  pandas/core/internals/blocks.py
pandas/core/internals/blocks.py:389: error: Unused "type: ignore" comment
pandas/core/internals/blocks.py:1142: error: Argument 1 to "np_can_hold_element" has incompatible type "Union[dtype[Any], ExtensionDtype]"; expected "dtype[Any]"  [arg-type]
Found 2 errors in 1 file (checked 1 source file)
```


```
> ./ci/code_checks.sh typing
mypy --version
mypy 0.930
Performing static analysis using mypy
pandas/core/internals/ops.py:128: error: No overload variant of "__getitem__" of "ExtensionArray" matches argument type "Tuple[int, slice]"  [call-overload]
...
pandas/core/internals/blocks.py:389: error: Unused "type: ignore" comment
pandas/core/internals/blocks.py:1142: error: Argument 1 to "np_can_hold_element" has incompatible type "Union[dtype[Any], ExtensionDtype]"; expected "dtype[Any]"  [arg-type]
...
Performing static analysis using mypy DONE

```

Should a new Issue about this be added? I would happily work on that

**Comment by mroeschke** [comment]
> ci checks are failing, however error seems unrelated
> 
> ```
> Run ci/code_checks.sh typing
> mypy --version
> mypy 0.931
> Performing static analysis using mypy
> pandas/core/internals/blocks.py:1142: error: Argument 1 to "np_can_hold_element" has incompatible type "Union[dtype[Any], ExtensionDtype]"; expected "dtype[Any]"  [arg-type]
> ```
> 
> Running mypy at the file confirm the error, and the CI yields an even more errors.
> 
> ```
> > mypy  pandas/core/internals/blocks.py
> pandas/core/internals/blocks.py:389: error: Unused "type: ignore" comment
> pandas/core/internals/blocks.py:1142: error: Argument 1 to "np_can_hold_element" has incompatible type "Union[dtype[Any], ExtensionDtype]"; expected "dtype[Any]"  [arg-type]
> Found 2 errors in 1 file (checked 1 source file)
> ```
> 
> ```
> > ./ci/code_checks.sh typing
> mypy --version
> mypy 0.930
> Performing static analysis using mypy
> pandas/core/internals/ops.py:128: error: No overload variant of "__getitem__" of "ExtensionArray" matches argument type "Tuple[int, slice]"  [call-overload]
> ...
> pandas/core/internals/blocks.py:389: error: Unused "type: ignore" comment
> pandas/core/internals/blocks.py:1142: error: Argument 1 to "np_can_hold_element" has incompatible type "Union[dtype[Any], ExtensionDtype]"; expected "dtype[Any]"  [arg-type]
> ...
> Performing static analysis using mypy DONE
> ```
> 
> Should a new Issue about this be added? I would happily work on that

Yup this is unrelated. Looks like it originated from https://github.com/pandas-dev/pandas/pull/45792 cc @jbrockmendel 
Would be happy to take a fix for it unless it's already being addressed in the meantime.

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks for all your work closing these regression tests @NumberPiOso! Happy to have more!

**Pull request merged**